cocotb/bfms.py

```diff
@@ -0,0 +1,206 @@
+    info.id = len(import_info_l)
+    import_info_l.append(info)
+
+def register_bfm_export_info(info):
+    info.id = len(export_info_l)
+    export_info_l.append(info)
+
+def bfm_hdl_path(py_file, template):
+    return os.path.join(
+        os.path.dirname(os.path.abspath(py_file)),
+        template)
+
+class BfmMethodParamInfo():
+    '''
+    Information about a single BFM-method parameter
+    '''
+
+    def __init__(self, pname, ptype):
+        self.pname = pname
+        self.ptype = ptype
+
+class BfmMethodInfo():
+    '''
+    Information about a single BFM method
+    - Method type
+    - User-specified parameter signature
+    '''
+
+    def __init__(self, T, signature):
+        fullname = T.__qualname__
+        fi = T.__code__
+
+        self.T = T
+        self.signature = []
+        self.type_info = []
+        self.id = -1
+
+        locals_idx = fullname.find("<locals>")
+        if locals_idx != -1:
+            fullname = fullname[locals_idx+len("<locals>."):]
+
+        if fullname.find('.') == -1:
+            raise Exception("Attempting to register a global method as a BFM method")
+
+        args = fi.co_varnames[1:fi.co_argcount]
+        if len(signature) != len(args):
+            raise Exception("Wrong number of parameter-type elements: expect " + str(len(args)) + " but received " + str(len(signature)))
+
+
+        for a,t in zip(args, signature):
+            self.signature.append(BfmMethodParamInfo(a, t))
+            try:
+                import simulator
+            except Exception:
+                # When we're not running in simulation, don't
+                # worry about being able to access constants from simulation
+                self.type_info.append(None)
+            else:
+                if isinstance(t, bfm_param_int_t):
+                    if t.s:
+                        self.type_info.append(simulator.BFM_SI_PARAM)
+                    else:
+                        self.type_info.append(simulator.BFM_UI_PARAM)
+
+class BfmTypeInfo():
+
+    def __init__(self, T, hdl, import_info, export_info):
+        self.T = T
+        self.hdl = hdl
+        self.import_info = import_info
+        self.export_info = export_info
+
+class BfmInfo():
+
+    def __init__(self, bfm, id, inst_name, type_info):
+        self.bfm = bfm
+        self.id = id
+        self.inst_name = inst_name
+        self.type_info = type_info
+
+    def call_method(self, method_id, params):
+        self.type_info.export_info[method_id].T(
+            self.bfm, *params)
+
+class BfmMgr():
+
+    _inst = None
+
+    def __init__(self):
+        self.bfm_l = []
+        self.bfm_type_info_m = {}
+        self.m_initialized = False
+
+    def add_type_info(self, T, type_info):
+        self.bfm_type_info_m[T] = type_info
+
+    @staticmethod
+    def get_bfms():
+        return BfmMgr.inst().bfm_l
+
+    @staticmethod
+    def find_bfm(path_pattern):
+        inst = BfmMgr.inst()
+        bfm = None
```

Comment:
I don't think these should be static methods - can you just make them regular methods, replace `inst` with `self`, and have the caller use `BgmMgr.inst().find_bfm(...)` instead?